### PR TITLE
feat: Const nodes are built with some extension requirements

### LINKED ISSUE #525: Extension inference for Const nodes

### PR DESCRIPTION
**Pull request by croyzor**
When using builder methods to put `Const` nodes on the graph, users must pass an extra "provenance" argument which specifies which extension the type of the constant value is coming from. The argument is an `Option`, where `None` means that the constant is of a built-in type (like our true and false predicates).

The `Const` nodes will be added to the graph with `input_extensions` which are either the empty set or the singleton set of the const value's provenance. Constant loading methods in the builder will make the `LoadConst` node's `input_extensions` match those of the constant it's connected too.

This also adds a constraint to extension inference that extension requirements are the same on both ends of `Static` and `ControlFlow` edges.

Resolves #525

### CODE REVIEW

File: src/extension/infer.rs [hunk 1]
```diff
@@ -323,10 +323,11 @@ impl UnificationContext {
...
             let sig: &OpType = hugr.get_nodetype(tgt_node).into();
             // Incoming ports with a dataflow edge
-            for port in hugr
-                .node_inputs(tgt_node)
-                .filter(|src_port| matches!(sig.port_kind(*src_port), Some(EdgeKind::Value(_))))
-            {
+            for port in hugr.node_inputs(tgt_node).filter(|src_port| {
+                matches!(sig.port_kind(*src_port), Some(EdgeKind::Value(_)))
+                    || matches!(sig.port_kind(*src_port), Some(EdgeKind::Static(_)))
+                    || matches!(sig.port_kind(*src_port), Some(EdgeKind::ControlFlow))
```
Comment (croyzor) [inline src/extension/infer.rs:329]: Adding `ControlFlow` edges as a generator of constraints is kind of irrelevant to this PR, just a drive-by

**Comment by acl-cqc** [comment]
Just a thought @croyzor but couldn't a constant require multiple extensions. For example, a List (that requires the collections extension) of Floats (requiring the arithmetic extension)....you get the idea.

Now, does an empty list of Floats require the arithmetic extension too? Ymmv....

Seems to me it ought to be possible to figure the extensions out from the `Value` rather than passing it separately, too.

File: src/extension/infer.rs (hunk 1 above)
Comment (acl-cqc) [reply src/extension/infer.rs:329]: I'd strongly suggest moving this into #529. It makes that more self-contained, more obviously complete (!), and it's really not relevant here

Comment (acl-cqc) [reply src/extension/infer.rs:329]: Also, can you do `matches!(sig.port_kind(*src_port), Some(EdgeKind::Value(_)) | Some(EdgeKind::Static(_)) | Some(EdgeKind:ControlFlow))`? That's possible in normal patterns (outside `matches!`), so maybe it works there too...

Comment (croyzor) [reply src/extension/infer.rs]: Removed the ControlFlow match

File: src/hugr/hugrmut.rs [hunk 2]
```diff
@@ -60,6 +60,14 @@ pub trait HugrMut: HugrView + HugrMutInternals {
         self.hugr_mut().add_op_before(sibling, op)
     }
 
+    /// A generalisation of [`HugrMut::add_op_before`], needed temporarily until
+    /// add_op type methods all default to creating nodes with open extensions.
```
Comment (cqc-alec) [inline src/hugr/hugrmut.rs:64]: Is there a github issue for this we could reference?

File: src/builder/build_traits.rs [hunk 3]
```diff
@@ -62,14 +62,21 @@ pub trait Container {
         Ok(Wire::new(src, src_port))
     }
 
-    /// Add a constant value to the container and return a handle to it.
+    /// Add a constant value to the container and return a handle to it. The
+    /// `provenance` argument says which extension the type and value of the
```
Comment (cqc-alec) [inline src/builder/build_traits.rs:66]: No `provenance` argument...

File: src/hugr/hugrmut.rs (hunk 2 above)
Comment (croyzor) [reply src/hugr/hugrmut.rs:64]: #424 - I'll add it to the comment

File: src/builder/build_traits.rs (hunk 3 above)
Comment (croyzor) [reply src/builder/build_traits.rs]: Removed

**Review by cqc-alec — APPROVED** [review]
